File: Archived/19단계/6549.py
import sys

def findMax(left,right):
    global fences
    
    if left == right:
        return fences[left]
    
    mid = (left + right)//2
    ret = max(findMax(left,mid),findMax(mid+1,right))
    l = mid
    r = mid+1
    currh = min(fences[l],fences[r])
    ret = max(ret, currh*2)
    while l>left or r<right:
        
        if l > left and (r == right or fences[r+1] < fences[l-1]):
            l -=1
            currh = min(currh, fences[l])
        else:
            r +=1
            currh = min(currh, fences[r])
        
        ret = max(ret, currh * (r-l+1))
    return ret
    
while 1:
    fence = sys.stdin.readline().strip()
    if fence == '0':
        break
    fences = list(map(int,fence.split()))[1:]
    ans = findMax(0,len(fences)-1)
    print(ans)

File: Archived/19단계/6549.java
import java.util.*;
import java.io.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static int[] tmp;
	public static void main(String[] args) throws IOException{
		tmp = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		long ans=0;
		while(tmp[0] != 0){
			ans = recur(1,tmp.length-1);
			System.out.println(ans);
			tmp = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
		}
		
	}
	static long recur(int left, int right){
		if(left == right)
			return tmp[left];
		int mid = (left+right)/2;
		long ret = Math.max(recur(left, mid), recur(mid+1, right));
		int l=mid, r=mid, minh=tmp[mid];
		while(l>left || r<right){
			if(r<right && (l==left || tmp[l-1]<tmp[r+1])){
				r++;
			}else{
				l--;
			}
			minh = Math.min(minh, Math.min(tmp[l],tmp[r]));
			ret = Math.max(ret, (long) (r - l + 1) *minh);
		}
		return ret;
	}
}